src/webgpu/shader/validation/shader_io/interpolate.spec.ts

```diff
@@ -0,0 +1,68 @@
+export const description = `Validation tests for the interpolate attribute`;
+
+import { makeTestGroup } from '../../../../common/framework/test_group.js';
+import { ShaderValidationTest } from '../shader_validation_test.js';
+
+import { generateShader } from './util.js';
+
+export const g = makeTestGroup(ShaderValidationTest);
+
+// List of valid interpolation attributes.
+const kValidInterpolationAttributes = [
+  '',
+  '[[interpolate(flat)]]',
+  '[[interpolate(perspective)]]',
+  '[[interpolate(perspective, center)]]',
+  '[[interpolate(perspective, centroid)]]',
+  '[[interpolate(perspective, sample)]]',
+  '[[interpolate(linear)]]',
+  '[[interpolate(linear, center)]]',
+  '[[interpolate(linear, centroid)]]',
+  '[[interpolate(linear, sample)]]',
+] as const;
+
+g.test('parameters')
+  .desc(`Test that all combinations of interpolation type and sampling are validated correctly.`)
+  .params(u =>
+    u
+      .combine('stage', ['vertex', 'fragment'] as const)
+      .combine('io', ['in', 'out'] as const)
+      .combine('use_struct', [true, false] as const)
+      .combine('type', ['', 'flat', 'perspective', 'linear'] as const)
+      .combine('sampling', ['', 'center', 'centroid', 'sample'] as const)
+      .beginSubcases()
+  )
+  .fn(t => {
+    if (t.params.stage === 'vertex' && t.params.use_struct === false) {
+      t.skip('vertex output must include a position builtin, so must use a struct');
+    }
+
+    let interpolate = '';
+    if (t.params.type !== '' || t.params.sampling !== '') {
+      interpolate = '[[interpolate(';
+      if (t.params.type !== '') {
+        interpolate += `${t.params.type}, `;
+      }
+      interpolate += `${t.params.sampling})]]`;
+    }
+    const code = generateShader({
+      attribute: '[[location(0)]]' + interpolate,
+      type: 'f32',
+      stage: t.params.stage,
+      io: t.params.io,
+      use_struct: t.params.use_struct,
+    });
+
+    t.expectCompileResult(
+      kValidInterpolationAttributes.some(x => x === interpolate),
```

Comment:
The CTS [uses `es2016`](https://github.com/gpuweb/cts/blob/59628599c4e4490228f1d7fdd4013be66d2a6946/tsconfig.json#L4) which has support for `Map` and `Set`s which would reduce the search complexity here.

Reply:
Done.